Jenkinsfile

```diff
@@ -7,120 +7,272 @@ pipeline {
-						])
-					}
+				dir ('eclipse.platform.swt') {
+					checkout scm
+					sh 'git fetch --all --tags --quiet'
+				}
+				dir ('eclipse.platform.swt.binaries') {
+					checkout([$class: 'GitSCM', branches: [[name: 'refs/heads/master']],
+						extensions: [[$class: 'CloneOption', timeout: 120, noTags: false ]],
+						userRemoteConfigs: [[url: 'https://github.com/eclipse-platform/eclipse.platform.swt.binaries.git']]
+					])
 				}
 			}
 		}
-		stage('Build') {
+		stage('Check if SWT-binaries build is needed') {
 			steps {
-				container('container') {
-					wrap([$class: 'Xvnc', useXauthority: true]) {
-						withEnv(["JAVA_HOME=${ tool 'openjdk-jdk17-latest' }"]) {
-							dir ('eclipse.platform.swt.binaries') {
-								sh '''
-									/opt/tools/apache-maven/latest/bin/mvn install \
-										--batch-mode -Pbuild-individual-bundles -DforceContextQualifier=zzz -Dnative=gtk.linux.x86_64 \
-										-Dcompare-version-with-baselines.skip=true -Dmaven.compiler.failOnWarning=true
-								'''
+				withAnt(installation: 'apache-ant-latest', jdk: 'openjdk-jdk11-latest') { // nashorn javascript-engine required in ant-scripts
+					sh'''
+						java -version
+						git config --global user.email 'eclipse-releng-bot@eclipse.org'
+						git config --global user.name 'Eclipse Releng Bot'
+						
+						ant -f eclipse.platform.swt/bundles/org.eclipse.swt/buildSWT.xml check_compilation -DTAG=${SWT_GIT_TAG}
+						ant -f eclipse.platform.swt/bundles/org.eclipse.swt/buildSWT.xml new_build_with_create_file -DTAG=${SWT_GIT_TAG}
+					'''
+				}
+			}
+		}
+		stage('Build SWT-binaries, if needed') {
+			when {
+				anyOf {
+					expression { return params.forceNativeBuilds }
+					expression { return fileExists ('tmp/build_changed.txt') && fileExists ('tmp/natives_changed.txt') }
+				}
+			}
+			matrix {
+				axes {
+					axis {
+						name 'PLATFORM'
+						values 'gtk.linux.aarch64', 'gtk.linux.ppc64le', 'win32.win32.x86_64' // 'cocoa.macosx.aarch64' , 'cocoa.macosx.x86_64', 'gtk.linux.aarch64', 'gtk.linux.ppc64le', 'gtk.linux.x86_64', 'win32.win32.x86_64'
+					}
+				}
+				stages {
+					stage("Collect SWT-native's sources") {
+						steps {
+							dir('eclipse.platform.swt.binaries/bundles'){
+								withAnt(installation: 'apache-ant-latest', jdk: 'openjdk-jdk11-latest') { // nashorn javascript-engine required in ant-scripts
+									sh '''
+										pfSpec=(${PLATFORM//"."/ })
+										ant -f binaries-parent/build.xml copy_library_src_and_create_zip -Dws=${pfSpec[0]} -Dos=${pfSpec[1]} -Darch=${pfSpec[2]}
+									'''
+								}
+								dir("org.eclipse.swt.${PLATFORM}/tmpdir") {
+									stash name:"swt.binaries.sources.${PLATFORM}",	includes: "org.eclipse.swt.${PLATFORM}.master.zip"
+								}
+							}
+						}
+					}
+					stage('Build SWT-native binaries') {
+						options {
+							skipDefaultCheckout()
+							timestamps()
+							timeout(time: 15, unit: 'MINUTES')
+						}
+						agent {
+							label "swt.natives-${PLATFORM}"
```

Comment:
Couldn't we have at the beginning of the file a map PLATFORM->agent with defined agents such as
```groovy
def agents = [ "gtk.linux.aarch64": agent { label "swt.natives.gtk.linux.aarch64" },
  ...
  "gtk.linux.x86_64": agent { label "centos-8" }]
```
and then let `agent` here just resolve the map with something like `agent[${PLATFORM}]`?

That would allow to have heterogeneous way of defining agents and to rely less on Eclipse infra convention. Later on, we could even consider replacing some agents defined with labels with some defined with a kubernetes config, typically if we want to use another linux before Eclipse infra sets it up.

Reply:
I tried that but failed to do it in a declarative pipeline, with errors like:
`WorkflowScript: 20: Method calls on objects not allowed outside "script" blocks.`
Therefore I think it is not possible (by design) to have such dynamic agents in a declarative pipeline.

But after thinking about this again and as the error message actually suggests, I had the idea to create the native-builds `node`s in scripted-pipeline style in the `script`-step I'm using for the build native stage anyway. In scripted style this should work.
This probably also enables to execute the third-stage of the matrix (to collect the build native-binaries) on the initial/pipeline agent again the (as I asked here https://stackoverflow.com/q/75201394/14542697).


> Later on, we could even consider replacing some agents defined with labels with some defined with a kubernetes config, typically if we want to use another linux before Eclipse infra sets it up.

In general I'm wondering if it is better to specify the infrastructure in the pipeline in this repo or if it's better to define it externally (e.g. in the jiro repo)? Initially I tended to the latter answer because external changes are often not that relevant (e.g. if a new version of cent-os is used or if the JIBB is restructured), but on the other hand it also allows specify more of the exact build environment used to build the natives, which is also good and allows to try out new environments in a verification build (as you said).

Nevertheless I'll try the scripted heterogeneous build-node definition now and if that works we can still decide to keep it or to proceed with https://github.com/eclipse-cbi/jiro/pull/261 and just use 'labeled' agents for all platforms.

Reply:
> Couldn't we have at the beginning of the file a map PLATFORM->agent with defined agents such as
> 
> ```groovy
> def agents = [ "gtk.linux.aarch64": agent { label "swt.natives.gtk.linux.aarch64" },
>   ...
>   "gtk.linux.x86_64": agent { label "centos-8" }]
> ```
> 
> and then let `agent` here just resolve the map with something like `agent[${PLATFORM}]`?
> 
> That would allow to have heterogeneous way of defining agents and to rely less on Eclipse infra convention. Later on, we could even consider replacing some agents defined with labels with some defined with a kubernetes config, typically if we want to use another linux before Eclipse infra sets it up.

All of that worked as hoped. 🎉 So if you prefer the current approach over https://github.com/eclipse-cbi/jiro/pull/261 it is possible. Instead of maintaining a map with closure-values I use a method that only handles the `linux.x86_64` case specially. This avoids the need to list each platform twice (once in the map and once in the matrix axis).

Reply:
Awesome and congrats for this! It looks like an advanced and useful solution of a usual pattern. I suspect it would be worth documenting somewhere (eg https://wiki.eclipse.org/Jenkins and/or some blog post if you use a blog).

IMO, it's better to get as much as possible of the infrastructure requirement specified in the project. It will make it easier for project to evolve its target infrastructure without having to involve EF infra foundation staff (who may not be available or busy with other things). I believe having the more infra defined in Jenkinsfile increases agility for the project.

Reply:
> Awesome and congrats for this! It looks like an advanced and useful solution of a usual pattern. I suspect it would be worth documenting somewhere (eg https://wiki.eclipse.org/Jenkins and/or some blog post if you use a blog).

Thanks. I don't have a blog, but documenting it in https://wiki.eclipse.org/Jenkins#FAQ sounds good. Looks like I don't have write permissions there. But if the infra-team is interested (@fredg02 ?) I can provide a little text with an example and one of the infra-team can adjust it as desired and add it to the side. If one of them write an FAQ entry all by themself I'm also fine with that. :)


> 
> IMO, it's better to get as much as possible of the infrastructure requirement specified in the project. It will make it easier for project to evolve its target infrastructure without having to involve EF infra foundation staff (who may not be available or busy with other things). I believe having the more infra defined in Jenkinsfile increases agility for the project.

That's right. But then the project also has to take care, for example of security updates etc. Not saying that is per-se a blocker, I understand and agree with your point, but I think the latter is something to consider as well.